Add Tensorflow with Horovod on MPI using Slurm

akantak opened the pull request:
Hi, we are working on a draft version of our Tensorflow with Horovod in Magpie. We were using it on Slurm with mpirun command.
Please take a look into this and let us know if there's something to improve.
Also, I know changing headers in auto-generated files is pointless, but these are our legal requirements (do you think it is a blocker?).

We know that there are still tests to be developed to be compliant with Magpie contribution guide. We will add them in the nearest future.

chu11 commented:
hi @akantak , looks really interesting.  There's a lot to review and I need to play with it, but in the meantime:

> Also, I know changing headers in auto-generated files is pointless, but these are our legal requirements (do you think it is a blocker?).

Shouldn't be a blocker, although perhaps I'll have to refactor the auto-generation so that some auto-generated files get > 1 header.

One question off the bat (and keep in mind I know nothing about horovod and very little of tensorflow), why is this feature specifically "tensorflow+horovod"?  My understanding is one can use tensorflow by itself, so perhaps there should be a autogenerated file for tensorflow and a different autogenerated file for tensorflow+horovod?  i.e. `magpie.sbatch-mpirun-tensorflow` and `magpie.sbatch-mpirun-tensorflow-horovod`?  Similar to how I have hadoop & hadoop+pig files.

I have not played with your PR yet, but I noticed the ```magpie-setup-projects-tensorflow-horovod``` is empty and the function ```Magpie_run_start_tensorflow_horovod()``` is basically empty.

My assumption is that `mpirun` & `horovod` sets up the appropriate tasks/processes necessary for the job to run?

But that begs the question, if `mpirun` handles all of the setup for the user, then why add this to magpie?  I think of Magpie's primary job to setup config files and start/teardown daemons that users may not necessarily be familiar with.  But if a simple `mpirun` call can setup tensorflow/horovod for the user, is a wrapper around it with Magpie necessary?

akantak commented:
> why is this feature specifically "tensorflow+horovod"

Based on our benchmarks suggest that TensorFlow performance is better when it is used with Horovod.

> if `mpirun` handles all of the setup for the user, then why add this to magpie?

Yes, mpirun can handle running the TensorFlow app but the setup is still bit involved process.  We feel Magpie should be extended to support TensorFlow. Here are a few reasons why...
- User needs to set a lot of parameters before running TensorFlow job every time. This is not an easy ‘mpirun’ in most cases. Magpie original objective was to make it easy to run Big Data frameworks on HPC clusters. It does a very job at it. But should we stop there? We don’t see new Big Data frameworks that needed to be supported by Magpie. However, Magpie can grow into supporting new deep learning frameworks. In doing so, we can make it easy to run Deep learning jobs for normal users.
- Most of the deep learning jobs will be preceded by a data prep job which is mostly done using Big Data frameworks such as spark. In most cases, TensorFlow lives in same clusters where Spark and other Big Data frameworks exist. Why not we extend Magpie to nicely package all in one place?
- There is constant growth in the number of artificial intelligence (Machine learning + Deep Learning + other) workloads that are landing on HPC infrastructures. We should consider extending Magpie objectives to fill this gap.
- Though TensorFlow leads in adoption, there are other deep learning frameworks like Caffe, Mxnet, Pytorch, Paddle Paddle, Keras etc. also been used widely. It most cases researchers try out multiple frameworks. Why not we reduce the learning curve required to run these frameworks by providing best practices in Magpie? 
- As you might have noticed, HPC+AI convergence is been a very important topic in the HPC community (Look at the number of talks at ISC & SC last couple of years). We feel Magpie can play important role in Retrofitting HPC infrastructures to support Big Data & AI workloads.

chu11 commented:
> Based on our benchmarks suggest that TensorFlow performance is better when it is used with Horovod.

I've read that there are other libraries out there that work on tensorflow.  Could there be other tensorflow+FOO sbatch files in the future?  Perhaps that would be a good reason to have a tensorflow sbatch file and a tensorflow+horovod sbatch file.

> User needs to set a lot of parameters before running TensorFlow job every time. 

Everything you mention sounds like a good argument for support in Magpie.  However, I don't see anything obvious that looks like tensorflow or horovod configuration in your PR.  Is that something you will be putting in the PR later?  That's the part that sort of confused me.  If Magpie configured a number of good defaults for tensorflow / horovod for the user, then the addition would make a lot more sense to me.

chu11 commented:
Additional question after talking to a co-worker.  Is it horovod that has mpirun support?  But tensorflow does not have mpirun support?

akantak commented:
Hey Al, sorry for late answer. Here're answers:

> Additional question after talking to a co-worker. Is it horovod that has mpirun support? But tensorflow does not have mpirun support?

You are right. It's horovod that is supporting mpi. Tensorflow supports MPI itself, but it does not scale that well.

> I've read that there are other libraries out there that work on tensorflow. Could there be other tensorflow+FOO sbatch files in the future? Perhaps that would be a good reason to have a tensorflow sbatch file and a tensorflow+horovod sbatch file.

Agree with you. Tensorflow is always changing. soon one may able to distribute TF without horovod. So we will as TF scripts.


> Everything you mention sounds like a good argument for support in Magpie. However, I don't see anything obvious that looks like tensorflow or horovod configuration in your PR. Is that something you will be putting in the PR later? That's the part that sort of confused me. If Magpie configured a number of good defaults for tensorflow / horovod for the user or set up some config files for the user, then the addition would make a lot more sense to me.

You are right, there is no explicit "configuration" of tensorflow or horovod, it could look a bit odd in contrast to different parts of Magpie, but every setting that is necessary is in MPI which is set up by slurm in this case. Our "added content" are instructions and examples of how to use slurm+MPI to easily run tensorflow with horovod on top of it. Also prepared two jobs:
- "synthetic benchmark" with sample code from the horovod repo, that works out-of-the-box, should be a good "starter" to make sure setup is properly prepared;
- "CNN benchmark" that is one of the tensorflow benchmarks, that have multiple predefined parameters and an instruction what to do to run it.

chu11 commented:
@akantak Thanks for the answers.  Just wondering where you are with your development on this?

After your PR came up, there is now interest internally about just having tensorflow support in Magpie and your tensorflow + horovod could be an extension.  Depending on how far along you are, I could work on the tensorflow part on a branch or we could start a branch with yours.

Keeping in mind I still know very little about tensorflow, my understanding is that tensorflow configs look like this on every python task that is run across the allocation:

```
cluster = tf.train.ClusterSpec({"JOBNAME": ["host1:2222", "host1:2223", "host2:2222", "host2:2223"]})
server = tf.train.Server(cluster, job_name="JOBNAME", task_index=MYRANK)
```

then callers can go from there.  My feeling is the hosts array & "MYRANK" could be exported from magpie via an environment variable and users could use that in their python scripts.  Perhaps some helper environment variables are needed like "MAX_TENSORFLOW_RANK", but that'd be it.

My initial test was going to be some trivial distributed addition such as the one documented by databricks:

https://databricks.com/tensorflow/distributed-computing-with-tensorflow

akantak commented:
@chu11 About this PR completeness, I thought that we would have to make changes after your review. We didn't plan to do more here besides tests. Let me know if it is OK, or if not, what you see that we need to add here.
I can add Tensorflow solo if we (you & I) would agree that I would do it. We just need to size the effort first.

chu11 commented:
I'm currently working on something.  When I finish up something prototype-ish, I'll push a branch and you can let me know what you think.

chu11 commented:
hey @akantak, I got an initial tree at a branch I call tensorflow.  https://github.com/LLNL/magpie/tree/tensorflow.  Conceptually, it's not all that different than what you have.  But there are some things I altered that I think will work out better for tensorflow+horovod and your work.

- I refactored a lot of the submission-script generation, including breaking out the copyright header.  So you would be able to generate your tensorflow+horovod submission files with the appropriate copyrights you need.

- submission-script generation also now has this notion of (what I am currently calling) "bigdata" vs "distributed" submission scripts.  Since we're running 1 task per node and that's all the work (vs starting daemons), lots of the original Magpie doesn't make sense.  This branch removes a lot of things from the submission scripts that wouldn't make sense for a tensorflow.  Configs such as "interactive mode" just don't make sense anymore.

  - this also makes it easier to generate templates for some schedulers/launchers and not others.

- I decided that slurm's `#SBATCH --no-kill` was a bad idea for this type of computing.  That's adjustable of course.

- I decided (possibly foolishly) in my branch that Magpie would run "python <user-provided-script>", where as you I think you have the user provide some script directly.  I'll have to think about that, but that's adjustable.

If you think this (at a high level, ignoring some details) is a pretty good point that you can add your tensorflow+horovod support, then I will clean it up a bit, add some tests, and push it into the magpie master.  Then we can iterate on your PR.

akantak commented:
Looks good, I'm currently working on merging correctly our branches (for now, branches are just merged, not working fine yet), I will add one more commit with fixes when it would be working.

chu11 commented:
great.  I'll send over some review of some nits I found in your original PR too.

akantak commented:
Hey @chu11, I've just made it work and removed some unnecessary (from my point of view) code. If I've deleted too much, let me know. Also, let me know what do you think.
Best regards

chu11 commented on NEWS:
```diff
@@ -1,3 +1,8 @@
+----------
+With sbatch & srun, add experimental support for distributed
+ Tensorflow and Tensorflow with Horovod on top of MPI support.
+
```
Instead, keep my original line, and add

`Add new submission script type "sbatchmpirun"`

Then

`With sbatch & mpirun, add experimental support for Tensorflow with Horovod`

chu11 commented on README.md:
```diff
@@ -2,11 +2,11 @@ Magpie
+supported.  It currently supports running over the parallel file
+system Lustre and running over any generic network filesytem.  There
+is scheduler/resource manager support for Slurm, Moab, Torque, and LSF.
 
```
Since it's not a tier 1 supported thing yet (and not with all scheduler + resource manager combos), lets just add a sentence "Experimental support for Tensorflow and Tensorflow with Horovod is also supported in some cases" after this paragraph.

chu11 commented on doc/README:
```diff
@@ -2,11 +2,12 @@ Magpie
+Zookeeper are supported.  It currently supports running over the
+parallel file system Lustre and running over any generic network
+filesytem.  There is scheduler/resource manager support for Slurm,
+Moab, Torque, and LSF.
```
similar adjustment to README.md

chu11 commented on magpie/run/magpie-run-project-tensorflow-horovod:
```diff
@@ -0,0 +1,74 @@
+    then
+        magpie_run_tensorflow_horovod_teardown_complete=1
+    fi
+}
```
for consistency to other code, I'd just have the line `magpie_run_tensorflow_horovod_teardown_complete=1` and nothing else.

chu11 commented on magpie/run/magpie-run-project-tensorflow-horovod:
```diff
@@ -0,0 +1,74 @@
+
+Magpie_run_start_tensorflow_horovod () {
+
+    if [ "${TENSORFLOW_HOROVOD_SETUP}" != "yes" ]
```
Your code is correct, but it took me a second to grok.  Could we stick to the code style
in the rest of the Magpie.  This pattern from tensorflow should be good.  Dunno if you want to output a small header for users?  Dunno if that'd be useful or not.

```
    if [ "${TENSORFLOW_SETUP}" == "yes" ] && [ "${magpie_run_prior_startup_successful}" == "true" ]
    then
        if [ "${MAGPIE_CLUSTER_NODERANK}" == "0" ]
        then
            echo "*******************************************************"
            echo "*"
            echo "* Tensorflow Information"
            echo "*"
            echo "* Running with python = ${MAGPIE_PYTHON}"
            echo "*******************************************************"
        fi

        magpie_run_tensorflow_should_be_torndown=1
        magpie_run_tensorflow_setup_successful=1
    else
        magpie_run_zookeeper_should_be_torndown=0
        magpie_run_zookeeper_setup_successful=1
    fi
....

akantak replied:
In quoted code, and also in code on the master branch, you are using zookeepers variables in 'else'. I will fix it in the commit.

chu11 replied:
whoops, thanks.  I will update in master.

chu11 commented on submission-scripts/script-templates/Makefile:
```diff
@@ -704,3 +760,39 @@ define create-storm
+define create-tensorflow-horovod
+	echo "Creating magpie.$(SCHED)-tensorflow-horovod"
+	cat 	magpie-shebang \
+		magpie-header-intel \
```
b/c it also includes llnl stuff, I think you need to include the llnl header too

akantak replied:
of course, my oversight, sorry

chu11 commented on submission-scripts/script-templates/magpie-magpie-customizations-substitution-tensorflow:
```diff
@@ -0,0 +1,115 @@
+############################################################################
```
I think you accidentally git added this file?  The "substitution" files shouldn't be added.

chu11 commented on submission-scripts/script-templates/magpie-tensorflow-horovod:
```diff
@@ -0,0 +1,90 @@
+
+# sourcing settings files, loading modules, activating virtualenvs should be placed here
+
+# custom MPIRUN_OPTIONS can be placed here
```
is this necessary?  The `magpie-run-job-mpirun-distributed` file already has an MPIRUN_OPTIONS.

chu11 commented:
i've pushed the tensorflow support to master.

akantak commented:
@chu11 Merged, fixed and addressed all review comments. Also fixed in https://github.com/LLNL/magpie/pull/303/files#diff-68bf32fda97eba1bde1fdc8a5730429c

akantak commented:
I would also add tests here.

akantak commented:
Hey @chu11 , I've added tests here, could you review?

chu11 commented on magpie-run-distributed:
```diff
@@ -33,46 +33,61 @@ source ${MAGPIE_SCRIPTS_HOME}/magpie/lib/magpie-lib-node-identification
-    echo "*"
-    echo "*******************************************************"
-fi
+echo "*******************************************************"
```
why the removal of the `if [ "${MAGPIE_CLUSTER_NODERANK}" == "0" ]`?  Wouldn't it output the same thing on each rank as a result?

akantak replied:
You are right, `if` should be there, I will fix it.

chu11 commented on testsuite/test-generate.sh:
```diff
@@ -606,6 +607,14 @@ if [ "${tensorflowtests}" == "y" ]; then
 fi
+if [ "${tensorflowhorovodtests}" == "y" ]; then
+    if [ "${standardtests}" == "y" ]; then
+        if [[ ${submissiontype} == *mpirun ]]
```
i don't think you need to check the submission type?  The `if [ -a ...` check in the test generation should be enough I think?

akantak replied:
You are right :) Removing it.

chu11 commented on doc/README.tensorflow-horovod.md:
```diff
@@ -0,0 +1,88 @@
+   * TensorFlow
+   * OpenMPI
+   * Horovod
+   * EPEL
```
I think you can remove EPEL, I think it's just assumed you are running on a system with an OS already :-)  I think OpenMPI doesn't have to be mentioned, as mpirun makes it obvious enough.

chu11 replied:
OR ... is a specific OpenMPI version required?  Perhaps it should be listed if that's a requirement.

akantak replied:
You are absolutely right, these things sounds pretty obvious here, they were copied one-to-one from our internal documentation where we weren't using Magpie, and it was just note about our setup. Removing obvious things.

chu11 commented:
Trying things out, noticed you need to add a 

```
tensorflowhorovodtests=n
```

near the top of `test-generate.sh` (we'll default to no since it's still experimental like tensorflow).

Just got it running:

```
Running benchmark...
Iter #0: 4.7 img/sec per CPU
```

it hasn't finished yet, but I think the fact it's running is enough.  Fix the little nit in the `test-generate.sh`, squash the commits into whatever combination you feel is cleaner set of commit messages, and rebase on master, and we can get this merged!

akantak commented:
@chu11 Fixed and squashed :)

chu11 commented:
awesome, thanks for this addition to Magpie.
